### PR TITLE
add http client to verify atlas token input against atlas cloud

### PR DESCRIPTION
**Pull request by ronenlu**
need to wait for atlas-cloud API to be ready.

https://github.com/ariga/gh-atlas/assets/63970571/cce01719-c467-49af-80eb-dc0a76046b5d

### CODE REVIEW

File: cloudpai/client.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	token string
+}
+
+// RoundTrip implements http.RoundTripper.
+func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header.Set("Authorization", "Bearer "+r.token)
+	req.Header.Set("User-Agent", "gh-atlas-cli")
+	req.Header.Set("Content-Type", "application/json")
+	return http.DefaultTransport.RoundTrip(req)
+}
```
Comment (masseelch) [inline cloudpai/client.go:31]: I prefer unexported types to come after exported. Just a matter of style though.

File: cloudpai/client.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+// roundTripper is a http.RoundTripper that adds the Authorization header.
+type roundTripper struct {
+	token string
+}
+
+// RoundTrip implements http.RoundTripper.
+func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header.Set("Authorization", "Bearer "+r.token)
+	req.Header.Set("User-Agent", "gh-atlas-cli")
```
Comment (masseelch) [inline cloudpai/client.go:28]: Maybe add a version?

File: cloudpai/client.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	req.Header.Set("Content-Type", "application/json")
+	return http.DefaultTransport.RoundTrip(req)
+}
+
+// Client is a client for the Atlas Cloud API.
+type Client struct {
+	client   *http.Client
+	endpoint string
+	token    string
```
Comment (masseelch) [inline cloudpai/client.go:37]: I think the Client does not need the token, does it? You pass it directly to the roundtripper in the constructor.

File: main.go [hunk 4]
```diff
@@ -8,6 +8,7 @@ import (
 	"log"
 	"math/rand"
 
+	"ariga.io/gh-atlas/cloudpai"
```
Comment (masseelch) [inline main.go:11]:
```suggestion
	"ariga.io/gh-atlas/cloudapi"
```

File: cloudpai/client.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+// roundTripper is a http.RoundTripper that adds the Authorization header.
+type roundTripper struct {
+	token string
+}
+
+// RoundTrip implements http.RoundTripper.
+func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header.Set("Authorization", "Bearer "+r.token)
+	req.Header.Set("User-Agent", "gh-atlas-cli")
+	req.Header.Set("Content-Type", "application/json")
```
Comment (masseelch) [inline cloudpai/client.go:29]: You set this both here and in the `post` method.

File: cloudpai/client.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	}
+	var scan = struct {
+		Data   any           `json:"data"`
+		Errors gqlerror.List `json:"errors,omitempty"`
+	}{
+		Data: data,
+	}
+	if err := json.NewDecoder(res.Body).Decode(&scan); err != nil && !errors.Is(err, io.EOF) {
+		return err
+	}
```
Comment (masseelch) [inline cloudpai/client.go:90]: Why don't you fail on EOF? This is still an error indicating parsing did not finish?

File: cloudpai/client.go (hunk 1 above)
Comment (ronenlu) [reply cloudpai/client.go:31]: roundTripper is not exported though

File: cloudpai/client.go (hunk 2 above)
Comment (ronenlu) [reply cloudpai/client.go:28]: I dont want to add build flags at the moment, as it is not necessary yet.

File: cloudpai/client.go (hunk 1 above)
Comment (masseelch) [reply cloudpai/client.go]: yes, this is why I like to have the code below the exported stuff. If I as a consumer visit the package I see what I can use first and don't have to scroll over loads of code I cannot access.

**Review by masseelch — APPROVED** [review]

**Pull request merged**